rasa/core/policies/ted_policy.py

```diff
@@ -654,264 +720,184 @@ def load(cls, path: Text) -> "TEDPolicy":
-            attention_dropout_rate=self.config[DROP_RATE_ATTENTION],
-            sparsity=self.config[WEIGHT_SPARSITY],
-            unidirectional=True,
-            use_key_relative_position=self.config[KEY_RELATIVE_ATTENTION],
-            use_value_relative_position=self.config[VALUE_RELATIVE_ATTENTION],
-            max_relative_position=self.config[MAX_RELATIVE_POSITION],
-            name=DIALOGUE + "_encoder",
-        )
-        self._tf_layers[f"embed.{DIALOGUE}"] = layers.Embed(
-            self.config[EMBEDDING_DIMENSION],
-            self.config[REGULARIZATION_CONSTANT],
-            DIALOGUE,
-            self.config[SIMILARITY_TYPE],
         )
-        self._tf_layers[f"embed.{LABEL}"] = layers.Embed(
-            self.config[EMBEDDING_DIMENSION],
-            self.config[REGULARIZATION_CONSTANT],
-            LABEL,
-            self.config[SIMILARITY_TYPE],
+        self._prepare_ffnn_layer(
+            LABEL, self.config[HIDDEN_LAYERS_SIZES][LABEL], self.config[DROP_RATE_LABEL]
         )
 
-    def _combine_sparse_dense_features(
-        self,
-        features: List[Union[np.ndarray, tf.Tensor, tf.SparseTensor]],
-        name: Text,
-        sparse_dropout: bool = False,
-    ) -> tf.Tensor:
+        self._prepare_transformer_layer(
+            DIALOGUE, self.config[DROP_RATE_DIALOGUE], self.config[DROP_RATE_ATTENTION]
+        )
 
-        dense_features = []
+        self._prepare_embed_layers(DIALOGUE)
+        self._prepare_embed_layers(LABEL)
 
-        for f in features:
-            if isinstance(f, tf.SparseTensor):
-                if sparse_dropout:
-                    _f = self._tf_layers[f"sparse_dropout.{name}"](f, self._training)
-                else:
-                    _f = f
-                dense_features.append(self._tf_layers[f"sparse_to_dense.{name}"](_f))
+    def _prepare_utterance_level_layers(self, name: Text) -> None:
+        for feature_type in POSSIBLE_FEATURE_TYPES:
+            if (
+                name not in self.data_signature
+                or feature_type not in self.data_signature[name]
+            ):
+                # features for feature type are not present
+                continue
+
+            self._prepare_sparse_dense_dropout_layers(
+                f"{name}_{feature_type}", self.config[DROP_RATE]
+            )
+
+            if name not in STATE_LEVEL_FEATURES:
+                self._prepare_sparse_dense_layers(
+                    self.data_signature[name][feature_type],
+                    f"{name}_{feature_type}",
+                    10,
+                )
             else:
-                dense_features.append(f)
+                self._prepare_sparse_dense_layers(
+                    self.data_signature[name][feature_type],
+                    f"{name}_{feature_type}",
+                    self.data_signature[name][feature_type][0][1],
+                )
+
+    def _prepare_encoding_layers(self, name: Text) -> None:
+        feature_type = SENTENCE
```

Comment:
what are these encoding layers? they're only dense sentence layers for `INTENT, TEXT, ACTION_NAME, ACTION_TEXT`

Reply:
`_prepare_utterance_level_layers` does not create any ffnn layers, it just creates layers for combine sparse and dense features

Reply:
ok..., but not only for user utterance also for bot

Reply:
we need comments 😭

Reply:
renamed the method and added a comment